agent/cache/cache.go

```diff
@@ -819,14 +847,22 @@ func (c *Cache) fetch(key string, r getOptions, allowNew bool, attempt uint, ign
 				return
 			}
 
+			var totalWait time.Duration
+
 			// If we're over the attempt minimum, start an exponential backoff.
 			if wait := backOffWait(attempt); wait > 0 {
-				time.Sleep(wait)
+				totalWait += wait
 			}
 
 			// If we have a timer, wait for it
 			if tEntry.Opts.RefreshTimer > 0 {
-				time.Sleep(tEntry.Opts.RefreshTimer)
+				totalWait += tEntry.Opts.RefreshTimer
+			}
+
+			select {
+			case <-time.After(totalWait):
```

Comment:
I find this a little subtle to read - it takes me a while to convince myself that the first case is not falling through to the `return` as well...

It's a nit as most go programmers are probably OK with this but I'd consider adding a comment to explicitly call out that the first case is just continuing execution. Or re-order the cases?